### PR TITLE
Unboxed types fixes extracted from PR1166

### PR DESCRIPTION
**Pull request by mshinwell**
- Fix to add return kind on `Project_value_slot` (plus necessary change in `Cmm_helpers`)
- Fixes for `Project_value_slot` in `Closure_conversion`
- Fix for arity check in `Simplify_apply_expr`
- Fixes for `More_type_creators`

Example from @chambart which now works:
```let f (x : unboxed_int64) =
  let () = Sys.opaque_identity () in
  let g (y : unboxed_int64) =
    let x' = (Int64.Unboxed.box x) in
    let y' = (Int64.Unboxed.box y) in
    let r = Int64.add x' y' in
    Int64.Unboxed.unbox r
  in
  let () = Sys.opaque_identity () in
  g```

### CODE REVIEW

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml [hunk 1]
```diff
@@ -592,12 +592,26 @@ let unary_primitive env res dbg f arg =
...
+      let memory_chunk : Cmm.memory_chunk =
+        match Flambda_kind.With_subkind.kind kind with
+        | Value -> Word_val
+        | Naked_number
+            (Naked_int32 | Naked_int64 | Naked_nativeint | Naked_immediate) ->
+          Word_int
+        | Naked_number Naked_float -> Double
+        | Region | Rec_info ->
+          Misc.fatal_errorf "Bad kind %a for Project_value_slot"
+            Flambda_kind.With_subkind.print kind
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_primitive.ml:609]: This code should probably be in a separate function in `to_cmm_shared` where other similar functions already exist (e.g. functions from flambda kinds to cmm machtypes).

Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_primitive.ml]: Fixed, and now also generates `Word_int` for tagged immediates.

**Comment by mshinwell** [comment]
@Gbury I've also improved the `To_cmm_shared` interface a bit to make better use of subkinds.  This actually seems simpler overall.

**Comment by mshinwell** [comment]
> @Gbury I've also improved the `To_cmm_shared` interface a bit to make better use of subkinds. This actually seems simpler overall.

Unfortunately this doesn't work because at Cmm level we lose the distinction between tagged integers and unboxed integers.  Reverting this changeset.

**Comment by Gbury** [comment]
> Unfortunately this doesn't work because at Cmm level we lose the distinction between tagged integers and unboxed integers. Reverting this changeset.

I don't understand what's the problem ?

**Comment by mshinwell** [comment]
> > Unfortunately this doesn't work because at Cmm level we lose the distinction between tagged integers and unboxed integers. Reverting this changeset.
> 
> I don't understand what's the problem ?

Will move this part to a separate PR to discuss

**Comment by mshinwell** [comment]
> > > Unfortunately this doesn't work because at Cmm level we lose the distinction between tagged integers and unboxed integers. Reverting this changeset.
> > 
> > 
> > I don't understand what's the problem ?
> 
> Will move this part to a separate PR to discuss

https://github.com/ocaml-flambda/flambda-backend/pull/1208

**Pull request merged**